src/Test/L1/Worker/L1TestBase.cs

```diff
@@ -0,0 +1,617 @@
+            LoadTasks();
+        }
+
+        private void LoadTasks()
+        {
+            var baseDirectory = Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location);
+            var taskZipsPath = Path.Join(baseDirectory, "TaskZips");
+            var tasksPath = Path.Join(baseDirectory, "L1", "Tasks");
+            if (!Directory.Exists(tasksPath))
+            {
+                throw new Exception("No mock tasks provided");
+            }
+            if (!Directory.Exists(taskZipsPath))
+            {
+                Directory.CreateDirectory(taskZipsPath);
+            }
+            foreach (var d in Directory.GetDirectories(tasksPath))
+            {
+                var zip = Path.Join(taskZipsPath, Path.GetFileName(d) + ".zip");
+                if (File.Exists(zip))
+                {
+                    File.Delete(zip);
+                }
+                ZipFile.CreateFromDirectory(d, zip);
+            }
+        }
+
+        private string GetLogFile(object testClass, string testMethod)
+        {
+            // Trim the test assembly's root namespace from the test class's full name.
+            var suiteName = testClass.GetType().FullName.Substring(
+                startIndex: typeof(Tests.TestHostContext).FullName.LastIndexOf(nameof(TestHostContext)));
+            var testName = testMethod.Replace(".", "_");
+
+            return Path.Combine(
+               Path.Combine(TestUtil.GetSrcPath(), "Test", "TestLogs"),
+               $"trace_{suiteName}_{testName}.log");
+        }
+
+        private async Task SetupMessage(HostContext context, Pipelines.AgentJobRequestMessage message)
+        {
+            // The agent assumes the server creates this
+            var jobServer = context.GetService<IJobServer>();
+            await jobServer.CreateTimelineAsync(message.Plan.ScopeIdentifier, message.Plan.PlanType, message.Plan.PlanId, message.Timeline.Id, default(CancellationToken));
+        }
+
+        private async Task<TestResults> RunWorker(HostContext HostContext, Pipelines.AgentJobRequestMessage message, CancellationToken jobRequestCancellationToken)
+        {
+            var worker = HostContext.GetService<IWorker>();
+
+            Task<int> workerTask = null;
+            // Setup the anonymous pipes to use for communication with the worker.
+            using (var processChannel = HostContext.CreateService<IProcessChannel>())
+            {
+                processChannel.StartServer(startProcess: (string pipeHandleOut, string pipeHandleIn) =>
+                {
+                    // Run the worker
+                    // Note: this happens on the same process as the test
+                    workerTask = worker.RunAsync(
+                        pipeIn: pipeHandleOut,
+                        pipeOut: pipeHandleIn);
+                }, disposeClient: false);
```

Comment:
I don't see where dispose is called. Seems like if you are setting disposeClient to false, you should call dispose somewhere here. If that's not needed, a comment should be here to explain why.

Reply:
So the reason for this is that our process is both the client and the server in this pipe. Normally, a process assumes only one role. `disposeClient` was added by me so that `ProcessChannel` wouldn't run `DisposeLocalCopyOfClientHandle()`. Disposing is handled by the `using` block and the `Dispose()` method of `ProcessChannel`.

I'll add a comment explaining this.